### PR TITLE
[SymmetricMemory] fix a bug where get_signal_pad() returns a tensor backed by a buffer ptr instead of a signal_pad ptr

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139677
* #139529
* #140129
* __->__ #140128
* #140127

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: test/distributed/test_symmetric_memory.py [hunk 1]
```diff
@@ -196,6 +196,9 @@ def test_get_signal_pad(self) -> None:
         symm_mem = _SymmetricMemory.rendezvous(t)
         peer_rank = (self.rank + 1) % self.world_size
 
+        signal_pad = symm_mem.get_signal_pad(self.rank)
```
Comment (lw) [inline test/distributed/test_symmetric_memory.py:199]: Some suggestions on making this test more robust:
- allocate a buffer of size zero, and then get its signal pad and write something to it (this is how I found this bug)
- write a signal to the signal pad, fill the buffer with zeros, check that the signal is still the same

**Review by lw — APPROVED** [review]
Great! Thanks!

File: test/distributed/test_symmetric_memory.py (hunk 1 above)
Comment (yifuwang) [reply test/distributed/test_symmetric_memory.py:202]: Appreciate the suggestion! Incorporated it into the test.

**Pull request closed without merging**